File: leetcode/Array/26-RemoveDuplicatesFromSortedArray/RemoveDuplicatesFromSortedArray1.cpp
#include <vector>
#include <iostream>

using namespace std;

// O(n) 用一个探路者索引去找不重复的元素，留一个基地索引用于存储不重复的元素。探路者只需O(1)即可找到所有不重复元素。
class Solution {
public:
    int removeDuplicates(vector<int>& nums) {
        if ( nums.size() == 0){
            return 0;
        }

        int base = 1;
        int explorer = 1;
        int length = nums.size();
        for ( ; explorer < length; explorer++ ){
            if ( nums[ explorer ] != nums[ explorer-1 ] ){
                nums[ base++ ] = nums[ explorer ];
            }
        }
        return base;
    }
};

int main( int argc, char** argv ){
    vector<int> nums = { 1, 1, 2, 2, 3, 3, 3, 5, 6, 8, 8, 20, 100 };
    Solution s;
    s.removeDuplicates( nums );
    for ( auto& i: nums ){
        cout << i << " ";
    }
}

File: leetcode/Array/0026-RemoveDuplicatesFromSortedArray/RemoveDuplicatesFromSortedArray1.py
class Solution:
    def removeDuplicates(self, nums):
        if len(nums) == 0:
            return 0
        
        base = 1
        explorer = 1
        length = len(nums)

        while explorer < length:
            if nums[ explorer ] != nums[ explorer-1 ]:
                nums[ base ] = nums [ explorer ]
                base+=1
            explorer+=1
        
        return base


nums = [1, 2, 3, 3, 3, 4, 4, 5, 70, 70, 72, 75, 100, 200, 200, 300]
s = Solution()
s.removeDuplicates( nums )
print( nums )